File: LeetCode/0513-find-bottom-left-tree-value/0513-find-bottom-left-tree-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findBottomLeftValue(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int value = 0;
        while (!q.isEmpty()) {
            int n = q.size();
            while (n > 0) {
                TreeNode temp = q.poll();
                value = temp.val;
                if (temp.right != null) {
                    q.offer(temp.right);
                }
                if (temp.left != null) {
                    q.offer(temp.left);
                }
                n -= 1;
            }
        }
        return value;
    }
}

// class Solution {
//     public int findBottomLeftValue(TreeNode root) {
//         Queue<TreeNode> q = new LinkedList<>();
//         q.offer(root);
//         int value = 0;
//         while (!q.isEmpty()) {
//             TreeNode temp = q.poll();
//             value = temp.val;
//             if (temp.right != null) {
//                 q.offer(temp.right);
//             }
//             if (temp.left != null) {
//                 q.offer(temp.left);
//             }
//         }
//         return value;
//     }
// }

// /**
//  * Definition for a binary tree node.
//  * public class TreeNode {
//  *     int val;
//  *     TreeNode left;
//  *     TreeNode right;
//  *     TreeNode() {}
//  *     TreeNode(int val) { this.val = val; }
//  *     TreeNode(int val, TreeNode left, TreeNode right) {
//  *         this.val = val;
//  *         this.left = left;
//  *         this.right = right;
//  *     }
//  * }
//  */
// class Solution {
//     public int value;
//     public int maxDepth;
//     public void solve(TreeNode root, int depth) {
//         if (root == null) {
//             return;
//         }
//         if (depth > maxDepth) {
//             maxDepth = depth;
//             value = root.val;
//         }
//         solve(root.left, depth + 1);
//         solve(root.right, depth + 1);
//     }
//     public int findBottomLeftValue(TreeNode root) {
//         maxDepth = -1;
//         solve(root, 0);
//         return value;
//     }
// }

File: LeetCode/0513-find-bottom-left-tree-value/0513-find-bottom-left-tree-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int value;
        while (!q.empty()) {
            int n = q.size();
            while (n--) {
                TreeNode* temp = q.front();
                value = temp->val;
                q.pop();
                if (temp->right) {
                    q.push(temp->right);
                }
                if (temp->left) {
                    q.push(temp->left);
                }
            }
        }
        return value;
    }
};

// class Solution {
// public:
//     int findBottomLeftValue(TreeNode* root) {
//         queue<TreeNode*> q;
//         q.push(root);
//         int value;
//         while (!q.empty()) {
//             TreeNode* temp = q.front();
//             value = temp->val;
//             q.pop();
//             if (temp->right) {
//                 q.push(temp->right);
//             }
//             if (temp->left) {
//                 q.push(temp->left);
//             }
//         }
//         return value;
//     }
// };

// /**
//  * Definition for a binary tree node.
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
//  * };
//  */
// class Solution {
// public:
//     int maxDepth;
//     int value;
//     void solve(TreeNode* root, int& depth) {
//         if (root == nullptr) {
//             return;
//         }
//         if (depth > maxDepth) {
//             maxDepth = depth;
//             value = root->val;
//         }
//         depth += 1;
//         solve(root->left, depth);
//         solve(root->right, depth);
//         depth -= 1;
//     }
//     int findBottomLeftValue(TreeNode* root) {
//         maxDepth = -1;
//         int depth;
//         solve(root, depth);
//         return value;
//     }
// };